RFC 49: Python 3 migration for wptserve (#49)
web-platform-tests/rfcs: rfcs/wptserve_py3.md

*   Custom handlers (test cases): 400+ custom handlers, ~20,000 lines \
This is a rough count of all Python files excluding some directories, which still includes some util scripts: `find . -name "*.py" \! -path "./tools/*" \! -path "./webdriver/*" \! -path "./docs/*" \! -path "./websockets/*" \! -path "./resources/*"`

## Goals

Roughly in the order of importance:

1.  Everything, including wptserve and custom handlers, must continue to work in both Python 2 and 3. (This was discussed and agreed upon at TPAC 2019.)
1.  Custom handlers must continue to be able to read and write raw bytes from requests and to responses, including headers and body. wptserve must not change the byte sequences in any way; no errors should be raised if the byte sequences do not have a valid encoding.
1.  _Optional/competing_: (Which goal to prioritize largely determines which option we choose.)
    a. Keep type semantics consistent across Python 2 and 3 (a string is either always binary or always text).
    a. Minimize code change to custom handlers, and consequently minimize the burden on test authors in the future by imposing fewer new requirements.

## Root problem (or; why is this hard?)

> In both Python 2 and 3, unprefixed string literals have the `str` type. However, the meaning of the `str` type has changed: in Python 2 it is a raw byte sequence, whereas in Python 3 it is a Unicode text string. The rationale behind the change was to move to a Unicode-by-default world: arguably most people who used `str` in Python 2 were probably dealing with text and should use Unicode, so this decision minimized the impact on code handling human-readable text. However in WPT we actually intended to use byte sequences (see [background](#background)).

Reviewer: Nit suggestion: s/WPT/wptserve and s/actually/often

```suggestion
In both Python 2 and 3, unprefixed string literals have the `str` type. However, the meaning of the `str` type has changed: in Python 2 it is a raw byte sequence, whereas in Python 3 it is a Unicode text string. The rationale behind the change was to move to a Unicode-by-default world: arguably most people who used `str` in Python 2 were probably dealing with text and should use Unicode, so this decision minimized the impact on code handling human-readable text. However in wptserve we often intended to use byte sequences (see [background](#background)).
```

This pain is further compounded by the choice of types in some parts of the Python 3 standard library. The low-level [`http`](https://docs.python.org/3/library/http.html) library takes and returns the `str` type (i.e. Unicode string) almost everywhere except the body (see the [type](https://github.com/python/typeshed/blob/master/stdlib/3/http/client.pyi) [annotations](https://github.com/python/typeshed/blob/master/stdlib/3/http/server.pyi)). This is convenient for existing code from Python 2 but semantically different and makes it difficult to send raw bytes on the wire. On the upside, the standard library carefully and intentionally uses an 8-bit isomorphic encoding (latin-1/ISO-8859-1) when appropriate (e.g. [headers](https://github.com/python/cpython/blob/8c3ab189ae552401581ecf0b260a96d80dcdae28/Lib/http/client.py#L220)) to allow arbitrary byte sequences:

*   Getting bytes out: “encode” the return value with latin-1
*   Passing bytes in: “decode” the byte sequence with latin-1

> ## Options

Reviewer: For clarity, I think we want to split these more clearly. This one should be something like `Proposed solution` (top-level, replacing 'Options'), and the other can move down to an Appendix or Alternatives with a note that it was not chosen (and why).

### [Preferred] Keep the semantics: always use byte sequences

The preferred approach of the community is to keep a consistent and semantically correct encoding model. That is to **always use byte sequences: `str` in Python 2 and `bytes` in Python 3** (with some rare exceptions below).

#### Detailed changes

wptserve changes:

*   Introduce a pair of helper functions `isomorphic_{encode,decode}` that use ISO-8859-1 under the hood and only encode/decode if necessary; both of them can accept either binary or text strings, but always return binary/text strings respectively regardless of the Python version.

> *   Most public APIs for custom handlers (including methods and properties of `Request` and `Response`) are binary, with the notable exception of the response body which can be text or binary strings.

Reviewer: Do we explain why 'response body' can be text or binary anywhere?

> *   Most public APIs for custom handlers (including methods and properties of `Request` and `Response`) are binary, with the notable exception of the response body which can be text or binary strings.

Reviewer: 'are binary' - should that be something like 'only accept and return binary'?

Author: ```suggestion
*   Most public APIs for custom handlers (including methods and properties of `Request` and `Response`) only accept and return binary, with the notable exception of the response body which can be text or binary strings (if a text string is given, it will be encoded with `Response.encoding` automatically).
```

> Guidelines for converting custom handlers:

Reviewer: ```suggestion
[Python file handlers](https://web-platform-tests.org/writing-tests/python-handlers/index.html?highlight=file%20handler):
```

> *   Make sure all strings are either always text or always bytes; all string literals in handlers should be prefixed with `b""` or `u""`. This will be enforced through a [lint](https://github.com/web-platform-tests/wpt/pull/23347). Existing code needs to be converted and reviewed manually ([example](https://github.com/web-platform-tests/wpt/pull/23363)). Note that we could add a hack to setters to accept both `str` and `bytes`, but on the getter side there is little we could do -- we don’t know whether the returned value will be used together with `bytes` or `str`, and things like `dict.get()` will be particularly tricky.

Reviewer: Suggest: change the first sentence slightly, and removed the ' Note that we could add a hack to setters to accept both `str` and `bytes`, but on the getter side there is little we could do -- we don’t know whether the returned value will be used together with `bytes` or `str`, and things like `dict.get()` will be particularly tricky.' because it doesn't seem relevant - we aren't adding this sort of helper, right?

```suggestion
*   All strings will be either always text or always bytes; all string literals in handlers should be prefixed with `b""` or `u""`.
    * This will be enforced through a [lint](https://github.com/web-platform-tests/wpt/pull/23347).
    * Existing code needs to be converted and reviewed manually ([example](https://github.com/web-platform-tests/wpt/pull/23363)).
```

*   **Headers** on both requests and responses are always *binary strings* for both keys and values. This includes:
    * **HTTP Basic Auth**: `Request.auth.{username,password}` are binary strings.
    * **Cookies**: Names and values in cookies are always binary strings, including `Request.cookies` and `Response.{set,unset,delete}_cookie`.

> *   **Request URL/form parameters** are all *binary strings* for both keys and values, accessible via `Request.GET` or `Request.POST`.

Reviewer: ```suggestion
*   **Request URL/form parameters** are always *binary strings* for both keys and values, accessible via `Request.GET` or `Request.POST`.
```

*   **Response body** can be written via `Response.writer.*` or via the [return values](https://web-platform-tests.org/tools/wptserve/docs/handlers.html#python-handlers), which can be either text or binary strings, but the two types should never be mixed and string literals must be prefixed.

>     * As an exception, `Response.set_error` takes only text message strings.

Reviewer: Can we briefly summarize why? (If not brief, then don't)

Author: ```suggestion
    * As an exception, `Response.set_error` takes only text message strings, as this message is used in the JSON response.
```
